core/raftlease/store.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raftlease
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/pubsub"
+	"github.com/juju/utils/clock"
+
+	"github.com/juju/juju/core/globalclock"
+	"github.com/juju/juju/core/lease"
+)
+
+// NotifyTarget defines methods needed to keep an external database
+// updated with who holds leases. (In non-test code the notify target
+// will generally be the state DB.)
+type NotifyTarget interface {
+	// Claimed will be called when a new lease has been claimed. Not
+	// allowed to return an error because this is purely advisory -
+	// the lease claim has still occurred, whether or not the callback
+	// succeeds.
+	Claimed(lease.Key, string)
+
+	// Expired will be called when an existing lease has expired. Not
+	// allowed to return an error because this is purely advisory.
+	Expired(lease.Key)
+
+	// Trapdoor returns a trapdoor to be attached to lease details for
+	// use by clients. This is intended to hold assertions that can be
+	// added to state transactions to ensure the lease is still held
+	// when the transaction is applied.
+	Trapdoor(lease.Key, string) lease.Trapdoor
+}
+
+// ReadonlyFSM defines the methods of the lease FSM the store can use
+// - any writes must go through the hub.
+type ReadonlyFSM interface {
+	Leases(time.Time) map[lease.Key]lease.Info
+	GlobalTime() time.Time
+}
+
+// StoreConfig holds resources and settings needed to run the Store.
+type StoreConfig struct {
+	FSM           ReadonlyFSM
+	Hub           *pubsub.StructuredHub
+	Target        NotifyTarget
+	RequestTopic  string
+	ResponseTopic func(requestID uint64) string
+
+	Clock          clock.Clock
+	ForwardTimeout time.Duration
+}
+
+// NewStore returns a core/lease.Store that manages leases in Raft.
+func NewStore(config StoreConfig) *Store {
+	return &Store{
+		fsm:      config.FSM,
+		hub:      config.Hub,
+		config:   config,
+		prevTime: config.FSM.GlobalTime(),
+	}
+}
+
+// Store manages a raft FSM and forwards writes through a pubsub hub.
+type Store struct {
+	fsm       ReadonlyFSM
+	hub       *pubsub.StructuredHub
+	requestID uint64
+	config    StoreConfig
+
+	prevTimeMu sync.Mutex
+	prevTime   time.Time
+}
+
+// ClaimLease is part of lease.Store.
+func (s *Store) ClaimLease(key lease.Key, req lease.Request) error {
+	err := s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationClaim,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+		Holder:    req.Holder,
+		Duration:  req.Duration,
+	})
+	if err == nil {
+		s.config.Target.Claimed(key, req.Holder)
+	}
+	return errors.Trace(err)
```

Comment:
I wonder if timeout should be handled differently?
What if system is heavily loaded and 5 seconds isn't enough? Should there be a retry or backoff and retry?

Reply:
The lease manager does retry with exponential backoff. I'm not totally sure about the numbers around this - I tried to do some experiments to get a feel for what they should be but it was pretty ad-hoc.